### PR TITLE
feat(benchmark): lineage-scoped promotion/demotion recommendations

### PR DESCRIPTION
**Pull request by jmoreira-valory**
> ⏳ **In progress.** The promote/demote **policy design doc** has moved to its own PR: #365. This PR carries the **code** for the first slice (lineage-scoped recommendations); the rest of the policy (edge-vs-market metric, paired-test gate, tournament / relevant-set) lands on later commits.

---

## TL;DR

Turns the daily benchmark's informational "promotion candidate" hint into an **explicit, gated `## Promotion / Demotion` recommendation** in the per-platform report + Slack post. Every comparison is **scoped to a tool's lineage** (`tool_lineage.json`), which is what makes demotion safe: the best version of each lineage can never be demoted, so the rule can never degenerate to "demote every tool" on hard markets.

## How a human reads it (sample output)

```
## Promotion / Demotion

**PROMOTE**
- 🟢 `factual_research-v2` (tournament Brier 0.1847, n=47) beats deployed
  `factual_research` 0.2832 by Δ -0.0985.

**DEMOTE**
- 🔴 `factual_research` — superseded by promotion candidate `factual_research-v2`.
- 🔴 `alpha` (Brier 0.30) — dominated by deployed sibling `alpha-v2` 0.20; redundant.
```

Quiet days (the common case) render a single line: `_No promotion or demotion candidates today._`

## Why scope by lineage

Tool variants have **different runtime names**, so a plain name-match never connects a candidate to the version it would replace. The lineage ledger does:

```
factual_research          ← deployed        (Brier 0.28)
└─ factual_research-v1
   └─ factual_research-v2  ← tournament cand (Brier 0.18)  ⇒ PROMOTE over factual_research
```

Decision flow:

```mermaid
flowchart TD
    C([Tournament candidate]) --> L{Deployed incumbent<br/>in the SAME lineage?}
    L -- none --> SKIP[no recommendation<br/>leave to callouts]
    L -- found --> G{Δ Brier ≥ 0.04?<br/>n ≥ 30 both sides?<br/>log-loss agrees?}
    G -- no --> NONE[no change]
    G -- yes --> P[🟢 PROMOTE candidate]
    P --> DM[🔴 DEMOTE the superseded incumbent]
    S([Two deployed versions,<br/>same lineage]) --> SD{worse beaten<br/>by ≥ 0.04?}
    SD -- yes --> DS[🔴 DEMOTE the redundant version]
```

The best version of a lineage is never on the losing side of either path, and different lineages never compare against each other ⇒ **no "demote everything".** A bad-but-best tool is an *improvement* target (the issue triage owns the absolute Brier floor), not a demotion target.

---

### Gates

Constants in `analyze.py`, beside the existing `CALLOUT_DELTA`:

| Constant | Value | Meaning |
|---|---|---|
| `PROMOTE_MIN_DELTA` | `0.04` | candidate must beat its lineage incumbent by this Brier margin — above the ~0.037 n=100 noise floor; mirrors `BRIER_REGRESSION_THRESHOLD` in the triage |
| `PROMOTE_MIN_N` | `30` | minimum resolved markets, on **both** candidate and incumbent |

- **PROMOTE** additionally requires log-loss to move the same direction as Brier (guards against a Brier-only fluke at small n). The incumbent is the **lowest-Brier deployed member of the candidate's lineage**.
- **DEMOTE** has no separate thresholds — it reuses `PROMOTE_MIN_DELTA` for the sibling-domination margin. Two paths, both within-lineage: (a) the incumbent a promoted candidate supersedes; (b) a deployed version dominated by a better deployed sibling.
- There is deliberately **no absolute level-floor demotion**. `tool_improvement_triage.py`'s `level_floor` trigger already owns "Brier above the no-skill baseline" and responds by opening an *improvement* issue — the correct response when there is no better alternative.

### Lineage resolution

- `load_tool_lineage()` reads `tool_lineage.json` into `{tool: parent}`; **fails open** to an empty map (every tool a singleton root) on any read/parse error, so a broken ledger degrades gracefully rather than hiding a recommendation.
- `_lineage_root()` walks parent links to the root, **cycle-safe**; tools absent from the ledger are their own root.

### Surfacing

- The section renders whenever tournament data is present (same guard as Tournament Callouts) and **always** prints a heading — either recommendations or the explicit "none today" line — so the daily summary confirms the criteria ran each day.
- `notify_slack.py` reproduces the PROMOTE/DEMOTE bullets verbatim (they are pre-formatted with the 🟢/🔴 prefix, labels, Brier, n, Δ).

### Scope / safety

- **Advisory only.** Nothing here deploys or registers a tool. The real promotion gate stays the human `agent-deployments` PR (gated by `verify_tool_hashes.py`) plus the PR-CI replay.
- All metrics are all-time, mirroring the Tournament Callouts section (tournament records are never day-windowed).

### Files

| File | Change |
|---|---|
| `benchmark/analyze.py` | lineage loader/root + deployed-member grouping + `_promotion_candidates` / `_demotion_candidates` / `section_promotion_demotion`, wired into `generate_report` |
| `benchmark/notify_slack.py` | Slack render rule for the new section |
| `benchmark/tests/test_analyze.py` | 14 promote/demote + 6 lineage-loader/root + 1 `generate_report` wiring test |

### Verification

314 benchmark tests pass; `black` / `isort` / `flake8` / `mypy` / `pylint` / `darglint` all green under the tomte configs.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by OjusWiZard — COMMENTED** [review]
Multi-perspective automated review (code review, comment accuracy, test coverage, silent-failure, type design, simplification). **No Critical bugs found** — the core safety invariant (best-of-lineage is never demoted) is correctly implemented and well tested. 5 Important findings and 7 advisories below as inline comments.

File: benchmark/analyze.py [hunk 1]
```diff
@@ -2242,6 +2244,23 @@ def section_version_deltas(
...
+# Tournament Callouts table. The recommendation is advisory only: the real
+# promotion gate stays the human ``agent-deployments`` PR plus the PR-CI
+# replay. Promotion and demotion are scoped to a tool's lineage (via
+# ``tool_lineage.json``) -- a tool is only ever compared against other
+# versions in its own lineage, never against an absolute Brier floor, so
+# the best version of each lineage can never be demoted (which would
+# otherwise degenerate to "demote every tool" on hard markets). There is
+# deliberately no absolute level-floor demotion: a bad-but-best tool is an
+# improvement target (the issue triage owns ``BRIER_LEVEL_THRESHOLD``), not
+# a demotion target. ``PROMOTE_MIN_DELTA`` sits above the n=100 Brier noise
```
Comment (OjusWiZard) [inline benchmark/analyze.py]: **Important — comment accuracy:** the justification cites the noise floor at the wrong sample size. The comment says `PROMOTE_MIN_DELTA` "sits above the **n=100** Brier noise floor (~0.037)", but the gate admits comparisons at `PROMOTE_MIN_N = 30`, where the noise floor is materially wider (roughly sqrt(100/30) ≈ 1.8×, i.e. ~0.067). As written, the comment claims a statistical guarantee the gate doesn't provide at its own minimum n. Either raise `PROMOTE_MIN_N`, or reword to acknowledge the n=30 floor is wider and 0.04 is a pragmatic compromise.

Also, "mirrors `BRIER_REGRESSION_THRESHOLD`" is loose: that threshold gates *regressions* on *rolling-window production* data; this one gates *improvements* on *all-time tournament* data. Same number, different semantics.

File: benchmark/analyze.py [hunk 2]
```diff
@@ -2550,6 +2569,321 @@ def section_tournament_callouts(
...
+    is missing or malformed, so a read error degrades to singleton
+    lineages rather than silently hiding a promotion or demotion.
+
+    :param path: path to ``tool_lineage.json``. Defaults to the ledger at
+        the repo root.
+    :return: ``{tool: parent_or_None}``; empty on any read/parse failure.
+    """
+    try:
+        raw = json.loads(path.read_text(encoding="utf-8"))
+    except (OSError, ValueError):
```
Comment (OjusWiZard) [inline benchmark/analyze.py]: **Important — silent failure, contradicts in-file precedent:** this swallow (and the `not isinstance(tools, dict)` branch below) emits no log, while the directly analogous fail-soft loader in this same file, `load_active_tournament_cids`, emits a `log.warning` precisely so the degraded state is attributable in run logs.

The docstring's "fails open … rather than silently hiding a promotion or demotion" is backwards for the case that matters: a corrupted/missing ledger collapses everything to singleton lineages, which **silently suppresses every cross-name promotion** (e.g. `factual_research-v2` vs `factual_research`) — exactly the feature this PR exists to provide. The operator just sees a plausible "no candidates today".

Suggested fix: mirror the `load_active_tournament_cids` warning:
```python
except (OSError, ValueError) as exc:
    log.warning(
        "Lineage scoping degraded to singleton lineages: could not load %s (%s)",
        path, exc,
    )
    return {}
```
and reword "fails open" to "degrades to singleton lineages, suppressing cross-name comparisons but never producing false recommendations".

File: benchmark/analyze.py [hunk 3]
```diff
@@ -2550,6 +2569,321 @@ def section_tournament_callouts(
...
+            continue
+        cand_label = release_map.resolve(cand_cid, rm)
+        delta = t_brier - inc_brier
+        promote_lines.append(
+            f"- 🟢 `{cand_tool}` `{cand_label}` (tournament Brier"
+            f" {t_brier:.4f}, n={valid_n}) beats deployed `{inc_tool}`"
+            f" `{inc_label}` {inc_brier:.4f} by Δ {delta:+.4f}."
+        )
+        # First candidate (Brier-sorted) to beat an incumbent wins it.
+        superseded.setdefault(inc_tool, (inc_label, inc_brier, cand_tool))
```
Comment (OjusWiZard) [inline benchmark/analyze.py]: **Important — duplicate PROMOTE bullets:** there is no per-`cand_tool` dedup in this loop. If a tool has two CIDs in `by_tool_version_mode` that are both in `active_cids` and both clear the gate (a real state: an older candidate still under evaluation alongside a newer one), two PROMOTE bullets are emitted against the same incumbent. `setdefault` keeps the DEMOTE side to one bullet, so the report reads "2 promotions, 1 demotion" for one decision.

Suggested fix: add a `promoted_tools: set[str]` guard mirroring the `seen` set in `_deployed_lineage_members` — or, if multiple bullets per tool are intended, document that explicitly and add a test asserting 2 PROMOTE + 1 DEMOTE (which would also pin that `setdefault` attributes the incumbent to the *stronger* candidate).

File: benchmark/analyze.py [hunk 4]
```diff
@@ -2550,6 +2569,321 @@ def section_tournament_callouts(
...
+        tournament_tvm, deployed_by_root, lineage, release_map_data, active_cids
+    )
+    demote_lines = _demotion_candidates(deployed_by_root, superseded)
+
+    lines = [
+        "## Promotion / Demotion",
+        "",
+        "_Actionable, lineage-scoped recommendations (advisory; rollout"
+        " stays a manual deployment PR). PROMOTE = a tournament candidate"
+        f" beats the deployed version of its lineage by Brier ≥"
```
Comment (OjusWiZard) [inline benchmark/analyze.py]: **Important — directionally ambiguous user-facing text:** in Brier scoring lower is better, so "beats the deployed version of its lineage by Brier ≥ 0.04" reads naturally as "has a Brier of at least 0.04" (the losing direction). The `_promotion_candidates` docstring gets it right ("Brier delta <= -PROMOTE_MIN_DELTA"); this rendered line is what humans see in the report/Slack. Suggest: "by a Brier **reduction** of ≥ 0.04".

File: benchmark/analyze.py [hunk 5]
```diff
@@ -2550,6 +2569,321 @@ def section_tournament_callouts(
...
+        ]
+        if not incumbents:
+            continue
+        inc_tool, _inc_cid, inc_brier, _inc_n, inc_ll, inc_label = min(
+            incumbents, key=lambda m: m[2]
+        )
+        if t_brier - inc_brier > -PROMOTE_MIN_DELTA:
+            continue
+        t_ll = t_stats.get("log_loss")
+        if t_ll is not None and inc_ll is not None and t_ll - inc_ll >= 0:
```
Comment (OjusWiZard) [inline benchmark/analyze.py]: **Advisory — boundary vs comment mismatch:** `>= 0` blocks promotion at an exact log-loss tie, but the comment on the next line says log-loss "disagrees". A tie is neither agreement nor disagreement. Blocking on tie is conservative and defensible, but either align the comment ("fails to improve") or change to `> 0` to match the stated intent. There's also no test at the tie boundary.

File: benchmark/analyze.py [hunk 6]
```diff
@@ -2550,6 +2569,321 @@ def section_tournament_callouts(
...
+    :return: ``(promote_lines, superseded)`` where ``superseded`` maps each
+        beaten deployed ``tool`` to ``(label, brier, candidate_tool)`` for
+        the demotion pass.
+    """
+    promote_lines: list[str] = []
+    superseded: dict[str, tuple[str, float, str]] = {}
+
+    # Best (lowest-Brier) qualifying candidate first, so bullet order is
+    # deterministic and the strongest case leads.
+    ordered = sorted(
```
Comment (OjusWiZard) [inline benchmark/analyze.py:2749]: **Advisory — non-deterministic ordering on Brier ties:** `section_tournament_callouts` uses a `(brier, tool)` tuple sort key for deterministic output; this sort uses bare Brier, so equal-Brier candidates render in dict-insertion order. Suggest appending `kv[0]` as a tiebreaker.

(Note: do **not** shortcut this to `kv[1].get("brier") or float("inf")` — a Brier of exactly `0.0` is falsy and would be sorted last. The existing `is not None` ternary is correct.)

File: benchmark/analyze.py [hunk 7]
```diff
@@ -2550,6 +2569,321 @@ def section_tournament_callouts(
...
+    """Walk parent links to the lineage root (cycle-safe).
+
+    :param tool: runtime tool name to resolve.
+    :param parents: ``{tool: parent_or_None}`` lineage map.
+    :return: the root ancestor's name -- ``tool`` itself when it has no
+        parent or is absent from the ledger.
+    """
+    seen: set[str] = set()
+    cur = tool
+    while cur not in seen:
```
Comment (OjusWiZard) [inline benchmark/analyze.py:2629]: **Advisory — silent cycle handling:** on a cyclic ledger this returns an arbitrary in-cycle node with no operator signal. A cycle can merge unrelated lineages under one synthetic root, producing wrong promote/demote pairings that the report reader has no reason to distrust. Suggest a `log.warning` when the loop exits via cycle detection (the fall-through `return cur` after the `while`).

File: benchmark/analyze.py [hunk 8]
```diff
@@ -2550,6 +2569,321 @@ def section_tournament_callouts(
...
+            continue
+        best = min(eligible, key=lambda m: m[2])
+        for tool, _cid, brier, valid_n, _ll, label in eligible:
+            if tool == best[0] or tool in demoted:
+                continue
+            if brier - best[2] >= PROMOTE_MIN_DELTA:
+                lines.append(
+                    f"- 🔴 `{tool}` `{label}` (Brier {brier:.4f},"
+                    f" n={valid_n}) — dominated by deployed sibling"
+                    f" `{best[0]}` `{best[5]}` {best[2]:.4f}"
```
Comment (OjusWiZard) [inline benchmark/analyze.py]: **Advisory — magic tuple indices:** the loop on the line above unpacks `_DeployedMember` by name, but `best` stays a raw tuple, so `best[0]`/`best[2]`/`best[5]` (tool/brier/label) are silent positional accesses — a field reorder in `_DeployedMember` would typecheck and produce wrong rendered output. Minimal fix, matching the `inc_*` unpack already used in `_promotion_candidates`:
```python
best_tool, _cid, best_brier, _n, _ll, best_label = min(eligible, key=lambda m: m[2])
```
Longer term, making `_DeployedMember` a `NamedTuple` would also clean up `m[1]`/`m[3]` in the incumbent filters and the `superseded` value triple.

File: benchmark/tests/test_analyze.py [hunk 9]
```diff
@@ -1761,6 +1764,242 @@ def test_sorted_by_brier_best_first(self) -> None:
...
+    def test_load_tool_lineage_fails_open_on_malformed_json(
+        self, tmp_path: Any
+    ) -> None:
+        """Malformed JSON fails open to an empty map."""
+        path = tmp_path / "tool_lineage.json"
+        path.write_text("not-json")
+        assert not load_tool_lineage(path)
+
+
+class TestPromotionDemotion:
```
Comment (OjusWiZard) [inline benchmark/tests/test_analyze.py:1848]: **Important — coverage gaps on the gate boundaries:** the suite covers the happy paths and single-gate blocks well, but misses:

1. **`active_cids=None`** — only the empty-set case is tested. This matters because `load_active_tournament_cids` returns `None` on a read failure, which disables CID scoping entirely; the exact failure path is untested.
2. **Exact `delta == -PROMOTE_MIN_DELTA` boundary** — current code promotes at exactly -0.04; a future `>` ↔ `>=` flip would pass the suite.
3. **Candidate CID == deployed CID** — the `m[1] != cand_cid` self-comparison guard has no test; removing it would create spurious self-promotions undetected.
4. **One-sided log-loss** — candidate has a terrible log-loss, incumbent has none → the guard intentionally doesn't fire; untested, so a future change to fire on one-sided data would go unnoticed.
5. **Two qualifying candidates in one lineage** — nothing asserts 2 PROMOTE + 1 DEMOTE, or that `setdefault` attributes the incumbent to the stronger candidate.

File: benchmark/tests/test_analyze.py [hunk 10]
```diff
@@ -1893,6 +2132,22 @@ def test_callout_section_included_when_candidate_present(self) -> None:
         )
         assert "## Tournament Callouts" in report
 
+    def test_promotion_demotion_section_wired_in(self) -> None:
```
Comment (OjusWiZard) [inline benchmark/tests/test_analyze.py:2206]: **Advisory — not hermetic:** unlike the `TestPromotionDemotion` tests (which pass `lineage={}`), this goes through `generate_report` and so calls the real `load_tool_lineage()` (repo-root `tool_lineage.json`) and `release_map.get_release_map()` (backed by `gh release list`, fail-soft). This matches the pre-existing callout wiring tests, so it's acceptable precedent — flagging mainly because if `tool_lineage.json` ever gains an entry colliding with `tool-a`, this test's `**PROMOTE**` assertion could change behavior.

File: benchmark/tests/test_analyze.py [hunk 11]
```diff
@@ -1761,6 +1764,242 @@ def test_sorted_by_brier_best_first(self) -> None:
         assert result.index("tool-best") < result.index("tool-worse")
 
 
+def _prod_cells(*cells: tuple) -> dict[str, Any]:
```
Comment (OjusWiZard) [inline benchmark/tests/test_analyze.py]: **Advisory — duplicated helper body:** `_prod_cells` and `_tourn_cells` share their entire body except the mode string and two constants. A shared `_score_cells(mode, da, bss, *cells)` with two thin wrappers would prevent silent divergence if a field is later added to one but not the other.

File: benchmark/notify_slack.py [hunk 12]
```diff
@@ -105,6 +105,8 @@ def _prompt_ranking_split(top_k: int) -> str:
 • promotion candidate: `superforcaster` (tournament `v0.20.0` vs production `v0.18.2`) — tournament Brier `0.1847` (n=47) vs production `0.2832`, Δ `-0.0985` improved
 • candidate: `factual_research` `untagged@bafybeib` — Brier `0.0896` (n=34), BSS vs mkt `+0.285` — no prod data
 
+*Promotion / Demotion:* REQUIRED — whenever the report contains a "## Promotion / Demotion" heading, output this section in full. Unlike the informational Tournament callouts above, these are gated, actionable verdicts. Reproduce every PROMOTE and DEMOTE bullet verbatim (they already carry the 🟢/🔴 prefix, tool + version labels in backticks, Brier, n, and Δ). Keep the PROMOTE list and DEMOTE list under their own *bold* sub-labels. If the section's only content is "No promotion or demotion candidates today.", output that single line so the reader sees the criteria ran. Never drop, merge, or re-rank this section.
```
Comment (OjusWiZard) [inline benchmark/notify_slack.py:108]: **Advisory — untested prompt addition:** no test asserts the prompt contains the new "Promotion / Demotion" directive. The prompt is a static string, so a one-line smoke test (e.g. `assert "Promotion / Demotion" in _prompt_ranking_split(5)`) would guard against accidental truncation in future edits.

File: benchmark/analyze.py (hunk 1 above)
Comment (jmoreira-valory) [reply benchmark/analyze.py]: Addressed in c5681448. Reworded: 0.04 is now described as a pragmatic compromise, explicitly **not** a significance guarantee — the noise floor is materially wider (~1.8x, ~0.067) at the gate's own `PROMOTE_MIN_N = 30`. Also noted the equality with `BRIER_REGRESSION_THRESHOLD` is coincidental: that gates *regressions* on *rolling-window production*, this gates *improvements* on *all-time tournament* — same number, different semantics. Kept `PROMOTE_MIN_N = 30` since the recommendation is explicitly advisory; the controlled candidate-vs-parent check is the downstream PR-CI cached replay.

File: benchmark/analyze.py (hunk 2 above)
Comment (jmoreira-valory) [reply benchmark/analyze.py]: Addressed in c5681448. Both fail-soft paths (the `except` and the `not isinstance(tools, dict)` branch) now emit a `log.warning` mirroring `load_active_tournament_cids`, naming the path and that cross-name promotions/demotions are suppressed. Docstring reworded: "degrades to singleton lineages … never produces a false recommendation, but suppresses cross-name comparisons, so the warning makes that state attributable in run logs."

File: benchmark/analyze.py (hunk 3 above)
Comment (jmoreira-valory) [reply benchmark/analyze.py]: Addressed in c5681448. Added a per-incumbent dedup: candidates are Brier-sorted, and the first (strongest) to clear the gate claims the incumbent via `if inc.tool in superseded: continue`, so a weaker rival — including another CID of the same tool — is skipped. The result is always 1 PROMOTE + 1 DEMOTE per decision. New `test_two_candidates_one_incumbent_single_promote_strongest` pins 1+1 and that the stronger candidate is the one attributed to the incumbent.

File: benchmark/analyze.py (hunk 4 above)
Comment (jmoreira-valory) [reply benchmark/analyze.py]: Addressed in c5681448 — the rendered blurb now reads "beats the deployed version of its lineage by a **Brier reduction** of >= 0.04".

File: benchmark/analyze.py (hunk 5 above)
Comment (jmoreira-valory) [reply benchmark/analyze.py]: Addressed in c5681448. Kept the conservative `>= 0` (blocks on an exact tie) and aligned the comment: "Brier improved but log-loss does not also improve (tie or worse) -> not a confirmed gain, treat as noise." Added `test_log_loss_tie_blocks_promotion` at the exact tie boundary.

File: benchmark/analyze.py (hunk 6 above)
Comment (jmoreira-valory) [reply benchmark/analyze.py:2749]: Addressed in c5681448 — added `kv[0]` (the tvm key) as the sort tiebreaker so equal-Brier candidates render deterministically. Kept the `is not None` ternary per your note, so a Brier of exactly 0.0 is not mis-sorted last.

File: benchmark/analyze.py (hunk 7 above)
Comment (jmoreira-valory) [reply benchmark/analyze.py:2629]: Addressed in c5681448 — `_lineage_root` now emits a `log.warning` on the cycle fall-through (after the `while`), naming the start tool and the synthetic root so the operator can distrust any promote/demote pairing this run produced for it.

File: benchmark/analyze.py (hunk 8 above)
Comment (jmoreira-valory) [reply benchmark/analyze.py]: Addressed in c5681448 — `_DeployedMember` is now a `NamedTuple`, so all access is by field name (`m.brier`, `m.cid`, `m.valid_n`, the incumbent `min(...)` result, and `best.tool`/`best.label`). No positional indexing remains in either the filters or the rendering.

File: benchmark/tests/test_analyze.py (hunk 9 above)
Comment (jmoreira-valory) [reply benchmark/tests/test_analyze.py:1848]: Addressed in c5681448 — added tests for all five:
1. `test_active_cids_none_disables_scoping`
2. `test_promote_at_exact_delta_boundary` (promotes at exactly -0.04, pinning `>` vs `>=`)
3. `test_candidate_cid_equal_to_deployed_cid_not_self_promoted` (the `m.cid != cand_cid` guard)
4. `test_one_sided_log_loss_does_not_block_promotion`
5. `test_log_loss_tie_blocks_promotion`, plus `test_two_candidates_one_incumbent_single_promote_strongest` for the 2-candidate dedup + stronger-attribution.

File: benchmark/tests/test_analyze.py (hunk 10 above)
Comment (jmoreira-valory) [reply benchmark/tests/test_analyze.py:2206]: `generate_report` doesn't plumb a lineage override, so it still goes through the real `load_tool_lineage()` (matching the pre-existing callout wiring tests, as you noted). To remove the collision risk you flagged, the test's tool is renamed to `wiretest-nonexistent-tool`, which can't appear in `tool_lineage.json`. c5681448.

File: benchmark/tests/test_analyze.py (hunk 11 above)
Comment (jmoreira-valory) [reply benchmark/tests/test_analyze.py]: Addressed in c5681448 — extracted `_score_cells(mode, da, bss, *cells)`; `_prod_cells` / `_tourn_cells` are now thin wrappers differing only in the mode string and the diagnostic stand-ins. (Docstring kept single-line in 49e2a6ca to satisfy darglint's short-strictness.)

File: benchmark/notify_slack.py (hunk 12 above)
Comment (jmoreira-valory) [reply benchmark/notify_slack.py:108]: Addressed in c5681448 — added `*Promotion / Demotion:*` to the `test_template_still_carries_core_sections` heading assertion, so an accidental truncation of the directive now fails the suite.

**Comment by LOCKhart07** [comment]
What about non tournament promotions. Tools that can be cache replayed fully?

**Comment by jmoreira-valory** [comment]
Closed in favour of #365
